### ISSUE TITLE
conn.c: segfault in connclose

### ISSUE THREAD

**Issue by Roguelazer** [beanstalkd/beanstalkd#328]
We're seeing an extremely rare segfault in 1.10. I don't have a core, but based on the EIP at the time of segfault, the error is happening at  https://github.com/kr/beanstalkd/blob/master/conn.c#L237 (`c->use` is null, so the deref is invalid and triggers a segfault).

I can't for the life of me figure out why `c->use` would be NULL. The only place it's explicitly set to NULL is on line 240, later in this same method.

This is the third time we've had beanstalkd segfault in the last year or so — it's just infrequent enough that I can't reproduce it by hammering beanstalkd with random data, but it's fairly catastrophic when it happens, so I also can't just ignore it and move on.

I am contemplating adding a guard (`if (c->use)`) before line 237 in our build, but that seems like it'd just be papering over the issue. I've seen this on multiple servers and multiple compiles of beanstalkd, so I'm fairly confident that it's not a cosmic ray or anything like that.

The segfault doesn't recur when beanstalkd restarts and loads the WAL, so it's doesn't seem to be a property of the data.

Here's what the logs end with:

```
Jul 30 05:51:13 $hostname beanstalkd[21487]: prot.c:1821 in update_conns: sockwant
Jul 30 05:51:13 $hostname beanstalkd[21487]: Operation not permitted
Jul 30 05:51:13 $hostname kernel: beanstalkd[21487]: segfault at 168 ip 0000000000401fa2 sp 00007ffde368d4b0 error 6 in beanstalkd[400000+e000]
```

It's all very strange.

**Comment by silviucpp** [comment]
@Roguelazer are you able to replicated this on the last master as well ?

Silviu

**Comment by Roguelazer** [comment]
I haven't even reprod it up our version yet. It will apparently take a lot of fuzzing...

-- 
James Brown,
currently mobile

**Comment by thorro** [comment]
It happens on a daily basis for us now, on v1.10 and v1.8. Strange that it worked flawlessly for 2 years. Changing the physical machine didn't help.

`Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00000000004021c0 in connsoonestjob (c=0x64d0f40) at conn.c:166
166	            if (j->r.deadline_at <= (soonest ? : j)->r.deadline_at) soonest = j;`

**Comment by ysmolsky** [comment]
It started to happen in our production as well. But it happens in libc?..
```
Oct  3 23:40:35 hostAA kernel: [18648946.623798] beanstalkd[16422]: segfault at 563a84345bb0 ip 00007f8320258d7c sp 00007fffe02ff498 error 6 in libc-2.17.so[7f8320107000+1b8000]
```
It seems to happen for particular kind of data but I could not catch it. Whenever I restart all the clients and restart beanstalkd then it disappear.

strace:
```
epoll_ctl(6, EPOLL_CTL_MOD, 13, {EPOLLPRI|EPOLLOUT|EPOLLRDHUP, {u32=45842328, u64=93862261129112}}) = 0
epoll_wait(6, [{EPOLLIN, {u32=32133728, u64=93862247420512}}], 1, 986) = 1
accept(3, {sa_family=AF_INET, sin_port=htons(48906), sin_addr=inet_addr("10.102.168.20")}, [16]) = 26
fcntl(26, F_GETFL)                      = 0x2 (flags O_RDWR)
fcntl(26, F_SETFL, O_RDWR|O_NONBLOCK)   = 0
epoll_ctl(6, EPOLL_CTL_ADD, 26, {EPOLLIN|EPOLLPRI|EPOLLRDHUP, {u32=45855976, u64=93862261142760}}) = 0
epoll_wait(6, [{EPOLLIN, {u32=45855176, u64=93862261141960}}], 1, 986) = 1
read(25, "use default\r\n", 208)        = 13
epoll_ctl(6, EPOLL_CTL_MOD, 25, {EPOLLPRI|EPOLLOUT|EPOLLRDHUP, {u32=45855176, u64=93862261141960}}) = 0
epoll_wait(6, [{EPOLLOUT, {u32=45854312, u64=93862261141096}}], 1, 986) = 1
write(24, "USING default\r\n", 15)      = 15
epoll_ctl(6, EPOLL_CTL_MOD, 24, {EPOLLIN|EPOLLPRI|EPOLLRDHUP, {u32=45854312, u64=93862261141096}}) = 0
epoll_wait(6, [{EPOLLIN, {u32=45853480, u64=93862261140264}}], 1, 986) = 1
read(23, "put 2147483648 0 120 10844531423"..., 208) = 208
--- SIGSEGV {si_signo=SIGSEGV, si_code=SEGV_MAPERR, si_addr=0x555e7c594130} ---
+++ killed by SIGSEGV +++
Segmentation fault
```

What tools can I use to investigate this?

**Comment by ysmolsky** [comment]
#160 is probably the same

**Comment by ysmolsky** [comment]
Unless we have a coredump from server compiled with -g, it's very hard to fix this issue or #160. I cannot reproduce this at all...

**Comment by ysmolsky** [comment]
Without a core dump it's next to impossible to fix this kind of error. If anyone can share core dump (that will contain everything that is inside of beanstalkd), I would be glad to look at and try to fix this. 
[How to get a core dump for a segfault on Linux](https://jvns.ca/blog/2018/04/28/debugging-a-segfault-on-linux/)
